org.scala-ide.sdt.core/src/org/scalaide/core/internal/project/scopes/BuildScopeUnit.scala

```diff
@@ -0,0 +1,90 @@
+import org.scalaide.core.internal.project.UnitBuilder
+import org.scalaide.ui.internal.preferences.ScalaPluginSettings
+import org.scalaide.util.internal.SettingConverterUtil
+import org.scalaide.core.internal.project.CompileScope
+import org.eclipse.core.resources.IMarker
+import org.eclipse.jdt.core.compiler.IProblem
+import org.eclipse.jdt.ui.text.java.IProblemLocation
+import org.eclipse.jdt.core.IJavaProject
+import org.eclipse.jdt.core.IJavaModelMarker
+
+case object CompileMacrosScope extends CompileScope {
+  def name = "Macros"
+}
+case object CompileMainScope extends CompileScope {
+  def name = "Main"
+}
+case object CompileTestsScope extends CompileScope {
+  def name = "Tests"
+}
+
+abstract class BuildScopeUnit(val owningProject: IScalaProject, val unitBuilder: UnitBuilder,
+    val dependentUnitInstances: Seq[BuildScopeUnit] = Seq.empty) {
+  @volatile private var error: Boolean = false
+
+  def isValidSourcePath(path: String): Boolean
+
+  def dependentUnits: Seq[CompileScope]
+
+  def scope: CompileScope
+
+  def hasError: Boolean = error
+
+  def compileUnit(addedOrUpdated: Set[IFile], removed: Set[IFile], monitor: SubMonitor): Unit = {
+    val getJavaSourcesOfThisScope: Unit => Set[IFile] = (Unit) => {
+      val Dot = 1
+      getToCompileFor(owningProject.allSourceFiles
+        .filter { _.getLocation.getFileExtension == SdtConstants.JavaFileExtn.drop(Dot) })
+    }
+    val resetJavaMarkers: Set[IFile] => Set[IFile] = javaFiles => {
+      javaFiles.foreach { _.deleteMarkers(IJavaModelMarker.JAVA_MODEL_PROBLEM_MARKER, true, IResource.DEPTH_INFINITE) }
+      javaFiles
+    }
+    val removeRemoved: Set[IFile] => Set[IFile] = _ -- removed
+    val resetJavaFiles = (getJavaSourcesOfThisScope andThen removeRemoved andThen resetJavaMarkers)()
+    error = !(areTransitiveDependenciesBuilt && areDependedUnitsBuilt &&
+      unitBuilder.buildUnit(getToCompileFor(addedOrUpdated ++ resetJavaFiles), getToCompileFor(removed), monitor, bulkOutputFilter))
+  }
+
+  private def areTransitiveDependenciesBuilt = {
+    val projectsInError = dependentUnits.flatMap { scope =>
+      owningProject.transitiveDependencies.filter(p => IScalaPlugin().getScalaProject(p).buildManager.hasErrors(scope))
+    }
+    val stopBuildOnErrorsProperty = SettingConverterUtil.convertNameToProperty(ScalaPluginSettings.stopBuildOnErrors.name)
+    val stopBuildOnErrors = owningProject.storage.getBoolean(stopBuildOnErrorsProperty)
+
+    if (stopBuildOnErrors && projectsInError.nonEmpty) {
+      owningProject.underlying.deleteMarkers(SdtConstants.ProblemMarkerId, true, IResource.DEPTH_INFINITE)
+      BuildProblemMarker.create(owningProject.underlying,
+        s"Project's ${scope.name} not built due to errors in dependent project(s) ${Set(projectsInError.map(_.getName)).mkString(", ")}")
+      false
+    } else true
+  }
+
+  private def areDependedUnitsBuilt = {
+    val wrongScopes = dependentUnitInstances filter { _.hasError } map { _.scope }
```

Comment:
Does this handle errors in upstream projects, or only errors in different scopes in the *same* project?